File: lab5/task1/java/Main.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Locale;
import java.util.Scanner;

public class Main {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        ArrayList<Product> productList = new ArrayList<>();

        try {
            while (true) {
                Product product = new Product();
                product.input(scanner);
                if (product.isExit())
                    break;
                productList.add(product);
            }

            String fileName = "products.txt";
            try (BufferedWriter writer = new BufferedWriter(
                    new OutputStreamWriter(new FileOutputStream(fileName), StandardCharsets.UTF_8))) {
                for (Product product : productList) {
                    product.diskOut(writer);
                }
            }
            System.out.println("\nДанные успешно записаны в файл " + fileName);

            ArrayList<Product> loadedProducts = new ArrayList<>();
            try (BufferedReader reader = new BufferedReader(
                    new InputStreamReader(new FileInputStream(fileName), StandardCharsets.UTF_8))) {

                String line;
                while ((line = reader.readLine()) != null) {
                    if (line.trim().isEmpty())
                        continue;
                    try {
                        Product product = new Product();
                        product.diskIn(line);
                        loadedProducts.add(product);
                    } catch (IllegalArgumentException e) {
                        System.err.println("Пропущен объект: " + e.getMessage());
                    }
                }
            }

            System.out.println("\n=== Считано из файла ===");
            for (Product product : loadedProducts) {
                product.output();
                System.out.println("Цена в долларах: " + product.toUSD(100));
                product.increaseIfToyota(100, 50);
                System.out.print("После проверки: ");
                product.output();
                System.out.println();
            }

        } catch (IOException e) {
            System.err.println("[Ошибка ввода/вывода]: " + e.getMessage());
        } catch (Exception e) {
            System.err.println("[Ошибка]: " + e.getMessage());
        }
    }
}

class Product {
    private String name;
    private double priceRub;
    private String manufacturer;

    public Product() {
    }

    public Product(String name, double priceRub, String manufacturer) {
        this.name = name;
        this.priceRub = priceRub;
        this.manufacturer = manufacturer;
    }

    public void input(Scanner scanner) {
        System.out.print("Введите наименование (0 - выход): ");
        name = scanner.nextLine();
        if (name.isEmpty())
            name = "0";
        if (name.equals("0"))
            return;

        System.out.print("Введите цену (в рублях): ");
        String priceStr = scanner.nextLine();
        if (priceStr.isEmpty())
            priceStr = "0";
        try {
            priceRub = Double.parseDouble(priceStr.replace(',', '.'));
        } catch (NumberFormatException e) {
            priceRub = 0;
        }

        System.out.print("Введите изготовителя: ");
        manufacturer = scanner.nextLine();
        if (manufacturer.isEmpty())
            manufacturer = "Неизвестный изготовитель";
    }

    public void output() {
        System.out.println("Товар: " + name + ", Цена: " + priceRub + " руб., Изготовитель: " + manufacturer);
    }

    public double toUSD(double rate) {
        return priceRub / rate;
    }

    public void increaseIfToyota(double rate, double increase) {
        if (name.toLowerCase(Locale.ROOT).contains("toyota"))
            priceRub += increase * rate;
    }

    public boolean isExit() {
        return name.equals("0");
    }

    public void diskOut(BufferedWriter writer) throws IOException {
        writer.write(name + ";" + priceRub + ";" + manufacturer);
        writer.newLine();
    }

    public void diskIn(String line) {
        String[] parts = line.split(";");
        if (parts.length < 3)
            throw new IllegalArgumentException("Неверный формат строки: " + line);

        name = parts[0].trim();

        try {
            priceRub = Double.parseDouble(parts[1].trim().replace(',', '.'));
        } catch (NumberFormatException e) {
            throw new IllegalArgumentException("Неверный формат цены: " + parts[1]);
        }

        manufacturer = parts[2].trim();
    }
}

File: lab5/task1/cpp/main.cpp
#include <iostream>
#include <vector>
#include <fstream>
#include <iostream>
#include <string>
#include <sstream>
using namespace std;

class Product
{
private:
    string name;
    double price_rub{};
    string manufacturer;

public:
    Product() = default;
    Product(string n, double p, string m)
        : name(std::move(n)), price_rub(p), manufacturer(std::move(m)) {}

    void input()
    {
        cout << "Введите наименование (0 - выход): ";
        getline(cin, name);
        if (name.empty())
            name = "0";
        if (name == "0")
            return;

        cout << "Введите цену (в рублях): ";
        string price_str;
        getline(cin, price_str);
        if (price_str.empty())
            price_str = "0";
        try
        {
            price_rub = stod(price_str);
        }
        catch (...)
        {
            price_rub = 0;
        }

        cout << "Введите изготовителя: ";
        getline(cin, manufacturer);
        if (manufacturer.empty())
            manufacturer = "Неизвестный изготовитель";
    }

    void output() const
    {
        cout << "Товар: " << name
             << ", Цена: " << price_rub
             << " руб., Изготовитель: " << manufacturer << endl;
    }

    double to_usd(double rate = 100.0) const { return price_rub / rate; }

    void increase_if_toyota(double rate = 100.0, double increase = 50.0)
    {
        if (to_lower_case(name).find("toyota") != string::npos)
            price_rub += increase * rate;
    }

    bool is_exit() const
    {
        return name == "0";
    }

    void disk_out(ofstream &out) const
    {
        if (!out)
            throw runtime_error("Ошибка открытия файла для записи");
        out << name << ';' << price_rub << ';' << manufacturer << '\n';
    }

    bool disk_in(ifstream &in)
    {
        string line;
        if (!getline(in, line))
            return false;

        if (line.empty())
            return disk_in(in);

        stringstream ss(line);
        string price_str;

        if (!getline(ss, name, ';'))
            return false;
        if (!getline(ss, price_str, ';'))
            return false;
        if (!getline(ss, manufacturer))
            manufacturer = "Неизвестный изготовитель";

        try
        {
            price_rub = stod(price_str);
        }
        catch (...)
        {
            price_rub = 0;
        }

        return true;
    }

    static string to_lower_case(string str)
    {
        for (char &c : str)
            c = static_cast<char>(tolower(static_cast<unsigned char>(c)));
        return str;
    }
};

int main()
{
    vector<Product> product_list;

    try
    {

        while (true)
        {
            Product product;
            product.input();
            if (product.is_exit())
                break;
            product_list.push_back(product);
        }

        {
            ofstream out("products.txt");
            if (!out.is_open())
                throw runtime_error("Не удалось открыть файл для записи");

            for (const auto &product : product_list)
                product.disk_out(out);

            cout << "\nДанные успешно записаны в файл products.txt\n";
        }

        vector<Product> loaded_products;
        {
            ifstream in("products.txt");
            if (!in.is_open())
                throw runtime_error("Не удалось открыть файл для чтения");

            Product product;
            while (product.disk_in(in))
            {
                loaded_products.push_back(product);
            }
        }

        cout << "\n=== Считано из файла ===\n";
        for (auto &product : loaded_products)
        {
            product.output();
            cout << "Цена в долларах: " << product.to_usd() << endl;
            product.increase_if_toyota();
            cout << "После проверки: ";
            product.output();
            cout << endl;
        }
    }
    catch (const exception &e)
    {
        cerr << "[Ошибка]: " << e.what() << endl;
        return 1;
    }

    return 0;
}
